[FIX] website: fix traceback on switching type of many2one form field

divy-odoo opened the pull request:
Steps to reproduce:

1. Drop the website_form snippet.
2. Select name field & change with the many2one field having no record.
3. Switch the type: "dropdown list" => "radio"

--> It will throw a traceback :

This commit, fix the uncaught promise by adding validation before using variable.

task-3932086

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/166108).

robodoo commented:
@divy-odoo the target branch 'saas-17.3' has been disabled, you may want to close this PR.

bvr-odoo reviewed (changes requested):
Hello @divy-odoo ,

As the issue happens only with fields that have no records in the database (for example, "alias domain" in the demo data) I think your fix is not the right one.

Fields with no records should not show in the "Field type" list. The records are hardcoded in the form, so it does not make sense to display fields with no values. This should also fix the bug at the same time.

The commit/PR message should also be updated for this change ;)

bso-odoo commented:
@robodoo delegate=bvr-odoo

msh-odoo commented:
> Hello @divy-odoo ,
> 
> As the issue happens only with fields that have no records in the database (for example, "alias domain" in the demo data) I think your fix is not the right one.
> 
> Fields with no records should not show in the "Field type" list. The records are hardcoded in the form, so it does not make sense to display fields with no values. This should also fix the bug at the same time.
> 
> The commit/PR message should also be updated for this change ;)

We should keep the old solution where we used optional chaining to avoid tracebacks.

Reason to Keep the Existing Solution:
If a field has already been dropped (e.g., dropped from the form snippet or all records in the relational model are deleted later on), the system should still function properly when the user changes the field type to "radio."

Concerns with Removing the Field from Options:
Your suggestion to remove the field from options won’t always work due to the following corner cases:

1. Dropped Form Snippet:
If a field with no records was already added to a form and later selected in edit mode, it will not work.

2. Records Deleted After Form Creation:
If a form is created when the relational model has records, but the user later deletes all records, what should happen in such cases?

Proposed Approach:
Users should not add fields that have no records.
Instead of adding complexity to the form logic, we can implement a warning when saving a form.
This warning will check if any relational fields added to the form have no records.

divy-odoo commented:
> > Hello @divy-odoo ,
> > As the issue happens only with fields that have no records in the database (for example, "alias domain" in the demo data) I think your fix is not the right one.
> > Fields with no records should not show in the "Field type" list. The records are hardcoded in the form, so it does not make sense to display fields with no values. This should also fix the bug at the same time.
> > The commit/PR message should also be updated for this change ;)
> 
> We should keep the old solution where we used optional chaining to avoid tracebacks.
> 
> Reason to Keep the Existing Solution: If a field has already been dropped (e.g., dropped from the form snippet or all records in the relational model are deleted later on), the system should still function properly when the user changes the field type to "radio."
> 
> Concerns with Removing the Field from Options: Your suggestion to remove the field from options won’t always work due to the following corner cases:
> 
> 1. Dropped Form Snippet:
>    If a field with no records was already added to a form and later selected in edit mode, it will not work.
> 2. Records Deleted After Form Creation:
>    If a form is created when the relational model has records, but the user later deletes all records, what should happen in such cases?
> 
> Proposed Approach: Users should not add fields that have no records. Instead of adding complexity to the form logic, we can implement a warning when saving a form. This warning will check if any relational fields added to the form have no records.

Hello @bvr-odoo, ping!

bvr-odoo commented:
Hello @divy-odoo @msh-odoo 
Sorry for the delay, I missed the ping.
Indeed, my suggestion was not right in cases where data are removed after the field has already been dropped with existing data. In this case, I would go back to your first proposal which is to handle it by verifying the existence of inputs where it has to be done. So keep it simple, we can keep the first commit of this PR.
IMO, a warning message is not necessary.

divy-odoo commented:
Hello @bvr-odoo, I have modified the PR with the first solution. thanks : ) 
Now PR is ready for the review & merge.

bvr-odoo commented:
> Hello @bvr-odoo, I have modified the PR with the first solution. thanks : ) Now PR is ready for the review & merge.

the PR still in DRAFT

divy-odoo commented:
Now, it's ready for the review. :smile:

bvr-odoo reviewed (changes requested):
Hello @divy-odoo  👋🏼 
While testing, I got a traceback with these steps:
- Create a "dropdown list" field for "alias domain"
- Set the visibility of "phone number" on the "alias domain" field
(Can you also add a test, please?)

![image](https://github.com/user-attachments/assets/f93d292e-a8f8-42ad-893e-3c44b1bbaaee)

divy-odoo commented:
Hello @bvr-odoo, Thanks for the noticing. 
I have fixed the traceback.  thanks : )

bvr-odoo reviewed (approved):
LGTM 👍🏼

bvr-odoo commented:
@robodoo r+
